### DISCUSSION TITLE
OpenDDS security connection issue when participant both publishes and subscribes to a topic

### DISCUSSION THREAD

**Discussion by sub131** [OpenDDS/OpenDDS#3751] [Q&A] [answered]
In an application with a single participant that is publishing a topic (app status) that the participant is also subscribing to, I'm getting authorization errors that prevents that connection (below). When security is disabled, it successfully connects and reads what it writes (a status reader that has it's own status message). Other applications, even if using the same keys, can connect in both directions.

Is this likely a configuration issue on my part, a bug, or is this a restriction of using DDS security that makes it behave different from unsecured applications? 

On a related note, why is the `Theserviceparticipant->get_security())` function missing when compiled without security? It seems like it would be more useful to include a version that always returns false. I was trying to use it to work around the above issue. I think I should be able to use an `#ifdef OPENDDS_SECURITY` and make my own `isSecured` function.

```
(2604|11620) WARNING: Sedp::generate_remote_matched_writer_crypto_handle() - Unable to lookup remote participant crypto info.
(2604|11620) WARNING: Sedp::create_datareader_crypto_tokens() - Unable to create local datareader crypto tokens with crypto key exchange plugin. Security Exception[-1.0]: Invalid remote writer handle
(2604|11620) WARNING: Sedp::generate_remote_matched_reader_crypto_handle() - Unable to lookup remote participant crypto info.
(2604|11620) WARNING: Sedp::create_datawriter_crypto_tokens() - Unable to create local datawriter crypto tokens with crypto key exchange plugin. Security Exception[-1.0]: Invalid remote reader handle
```

**Comment by iguessthislldo** [comment, marked as answer]
There is a test for this called [SingleParticipantWithSecurity](https://github.com/objectcomputing/OpenDDS/tree/master/tests/security/SingleParticipantWithSecurity). I would suggest building and running the test as a first step to make sure nothing is funny about your build. If the test doesn't fail, it probably is a configuration related issue as the test has bare minimum security settings. As you suggested this could be a misconfiguration on your part or we're not testing a case we could in the test and there's a bug.

If your governance and permissions files differ from the ones in the test, you should post them here. Alternatively if you're sure it's not misconfiguration and you're feeling up for it, you could extend the test with your governance and permissions settings. If it fails in the same way, you can post it as a PR. That would help us get a start on fixing the bug and make sure it says fixed.

Also I have had that same idea about `get_security` before, but never got around to changing it.

**Comment by mitza-oci** [comment]
@sub131 could you describe a bit more how you intend to use the `get_security()` function?  What code in your application would be conditional on the return value from it?  Could that code run unconditionally (even if it essentially has no effect when DDS Security is disabled)?
